### PR TITLE
Move py installation out of the narr-base-image

### PR DESCRIPTION
**Pull request by ialarmedalien**
Also taking the liberty of update bullseye to bookworm and from py3.10 to py3.12.

Closes #82

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -21,6 +23,7 @@ RUN mkdir -p /kb/deployment/bin && \
     # for R and CRAN installations
     libxml2 \
     libxml2-dev \
+    libxslt1-dev \
```
Comment (ialarmedalien) [inline Dockerfile:26]: needed when updating py 3.10 -> 3.12

File: Dockerfile [hunk 2]
```diff
@@ -33,6 +36,11 @@ RUN mkdir -p /kb/deployment/bin && \
     glpk-doc \
     cmake \
     r-base r-base-dev && \
+    # Install Dockerize
+    curl -LJO https://github.com/kbase/dockerize/raw/master/dockerize-linux-amd64-v0.6.1.tar.gz && \
+    tar xvzf dockerize-linux-amd64-v0.6.1.tar.gz && \
+    mv dockerize /kb/deployment/bin && \
+    rm dockerize-linux-amd64-v0.6.1.tar.gz && \
     # JavaScript needs
```
Comment (ialarmedalien) [inline Dockerfile:44]: move up Dockerize install

File: Dockerfile [hunk 3]
```diff
@@ -41,36 +49,16 @@ RUN mkdir -p /kb/deployment/bin && \
     rm -rf /var/lib/apt/lists/* /var/cache/apt/archives/*
 
 # Install Base libraries, Node, R and Jupyter Notebook and ipywidgets from distinct channels
-ADD ./ /root/image_files
+ADD ./install-requirements/install-r-packages.R /root/image_files/install-r-packages.R
 
 # R requirements
 RUN R --vanilla < /root/image_files/install-r-packages.R
 
-# Python dependencies
```
Comment (ialarmedalien) [inline Dockerfile:49]: this will all happen in the narrative image from now on. Brave new world!

File: install-r-packages.R [hunk 4]
```diff
@@ -4,10 +4,6 @@ install.packages("devtools")
 install.packages("fACD", repos="http://R-Forge.R-project.org")  # ACD model
 install.packages("Rglpk", type="source")
 install.packages("Quandl", type="source")
-install.packages("pkgconfig")
-install.packages("knitr")
-install.packages("plyr")
```
Comment (ialarmedalien) [inline install-r-packages.R:9]: not important enough to be installed twice

File: .github/workflows/build_and_test_pr.yml [hunk 5]
```diff
@@ -1,29 +1,29 @@
...
-      GHCR_USERNAME: ${{ secrets.GHCR_USERNAME }}
-      GHCR_TOKEN: ${{ secrets.GHCR_TOKEN }}
+    run_build_and_push:
+        uses: ./.github/workflows/build_pr_image.yml
+        secrets:
+            GHCR_USERNAME: ${{ secrets.GHCR_USERNAME }}
+            GHCR_TOKEN: ${{ secrets.GHCR_TOKEN }}
 
-  trivy_scan:
-    if: (github.base_ref == 'develop' || github.base_ref == 'main' || github.base_ref == 'master' ) && github.event.pull_request.merged == false
```
Comment (ialarmedalien) [inline .github/workflows/build_and_test_pr.yml:20]: this repo uses "main"

File: .github/workflows/test_pr_image.yml [hunk 6]
```diff
@@ -2,66 +2,58 @@
...
-      id: test_frontend
-      shell: bash -l {0}
-      run: |
-        docker image ls -a
-        bash scripts/run_tests.sh -u -c ${{ env.NARRATIVE_IMG_TAG }}
-      continue-on-error: true
+    build_narrative_image:
+        runs-on: ubuntu-latest
+        steps:
+            - name: Pull Narrative Repo
```
Comment (ialarmedalien) [inline .github/workflows/test_pr_image.yml:18]: VSCode reformatted all this code, rather annoyingly. Apologies!

File: .github/workflows/test_pr_image.yml [hunk 7]
```diff
@@ -2,66 +2,58 @@
...
+              with:
+                  context: .
+                  file: ./Dockerfile
+                  pull: true
+                  load: true
+                  tags: ${{ env.NARRATIVE_IMG_TAG }}
+            - name: Use Node JS
+              uses: actions/setup-node@v2
+              with:
+                  node-version: 20
```
Comment (ialarmedalien) [inline .github/workflows/test_pr_image.yml:38]: this is the only change in this file -- update node version to 20 (replicates the change in #82 )

File: Dockerfile (hunk 1 above)
Comment (ialarmedalien) [inline Dockerfile:26]: not present by default in the slim-bookworm py3.12 image

File: Dockerfile [hunk 8]
```diff
@@ -33,6 +36,11 @@ RUN mkdir -p /kb/deployment/bin && \
     glpk-doc \
     cmake \
     r-base r-base-dev && \
+    # Install Dockerize
```
Comment (ialarmedalien) [inline Dockerfile:39]: move up Dockerize install

File: Dockerfile [hunk 9]
```diff
@@ -1,8 +1,10 @@
-FROM python:3.10-slim-bullseye
+FROM python:3.12-slim-bookworm
+
+# These ARGs values are passed in via the docker build command
+ARG BUILD_DATE
+ARG VCS_REF
+ARG BRANCH=main
```
Comment (ialarmedalien) [inline Dockerfile:6]: these args are referenced in the LABEL command

File: install-r-packages.R [hunk 10]
```diff
@@ -1,39 +1,73 @@
+# This script attempts to install all the R dependencies for the narrative base image.
+# It goes through all the packages listed, attempts to install them, and saves an error
+# message if the install fails.
```
Comment (ialarmedalien) [inline install-r-packages.R:3]: Rejigged the install script a bit (thanks chatGPT!) - it tries to install everything and collects errors as it goes along. If there are failures, the error messages are listed and the install script exits with a non-zero status.

**Comment by briehl** [comment]
I'm a little nervous of updating to 3.12, since I read about packages breaking under that all the time. But I guess if tests pass?

File: .github/workflows/test_pr_image.yml (hunk 7 above)
Comment (briehl) [reply .github/workflows/test_pr_image.yml:38]: Can we close #82 then?

**Review by briehl — CHANGES_REQUESTED** [review]
Again, another chicken-and-egg moment. I think this is fine, we migrate all the Python dependencies to the `narrative` repo and image. But I think this might not build and test right until we do.

(I think we're talking about this on Slack, just recording here for posterity).

Anyway, I think we can bump to 8.0.0 in the release notes, and this'll be fine.

File: .github/workflows/test_pr_image.yml (hunk 7 above)
Comment (ialarmedalien) [reply .github/workflows/test_pr_image.yml:38]: Yes, I'll add a tag to the description so it happens automatically.

**Review by briehl — APPROVED** [review]

**Pull request merged**